File: src/main/java/com/stratio/connector/hdfs/engine/HDFSMetadataEngine.java
package com.stratio.connector.hdfs.engine;

import java.util.List;
import java.util.Map;

import com.stratio.connector.commons.connection.Connection;
import com.stratio.connector.commons.engine.CommonsMetadataEngine;
import com.stratio.connector.hdfs.connection.HDFSConnectionHandler;
import com.stratio.connector.hdfs.utils.HDFSClient;
import com.stratio.crossdata.common.data.CatalogName;
import com.stratio.crossdata.common.data.ClusterName;
import com.stratio.crossdata.common.data.TableName;
import com.stratio.crossdata.common.exceptions.ConnectorException;
import com.stratio.crossdata.common.exceptions.ExecutionException;
import com.stratio.crossdata.common.exceptions.UnsupportedException;
import com.stratio.crossdata.common.metadata.CatalogMetadata;
import com.stratio.crossdata.common.metadata.IndexMetadata;
import com.stratio.crossdata.common.metadata.TableMetadata;
import com.stratio.crossdata.common.statements.structures.Selector;

public class HDFSMetadataEngine extends CommonsMetadataEngine<HDFSClient>{


    public HDFSMetadataEngine(HDFSConnectionHandler connectionHandler) {
        super(connectionHandler);
    }



    @Override
    protected void createCatalog(CatalogMetadata catalogMetadata, Connection<HDFSClient> connection)
            throws UnsupportedException, ExecutionException {

        connection.getNativeConnection().mkdir(catalogMetadata.getName().getName());
    }

    @Override
    protected void createTable(TableMetadata tableMetadata, Connection<HDFSClient> connection)
            throws UnsupportedException, ExecutionException {
        connection.getNativeConnection().addFile(tableMetadata.getName().getCatalogName().getName()+"/"+
                tableMetadata.getName().getName());
        connection.getNativeConnection().createMetaDataFile(tableMetadata);

    }

    @Override
    protected void dropCatalog(CatalogName catalogName, Connection<HDFSClient> connection)
            throws UnsupportedException, ExecutionException {

        connection.getNativeConnection().deleteFile(catalogName.getName());

    }

    @Override
    protected void dropTable(TableName tableName, Connection<HDFSClient> connection)
            throws UnsupportedException, ExecutionException {

        connection.getNativeConnection().deleteFile(tableName.getCatalogName().getName()+"/"+
                tableName.getName());
    }

    @Override
    protected void createIndex(IndexMetadata indexMetadata, Connection connection)
            throws UnsupportedException, ExecutionException {

        throw new UnsupportedException("Not yet supported");
    }

    @Override
    protected void dropIndex(IndexMetadata indexMetadata, Connection connection)
            throws UnsupportedException, ExecutionException {

        throw new UnsupportedException("Not yet supported");
    }

    @Override protected List<CatalogMetadata> provideMetadata(ClusterName clusterName,
            Connection<HDFSClient> connection) throws ConnectorException {
        throw new UnsupportedException("Not yet supported");
    }

    @Override protected CatalogMetadata provideCatalogMetadata(CatalogName catalogName, ClusterName clusterName,
            Connection<HDFSClient> connection) throws ConnectorException {
        throw new UnsupportedException("Not yet supported");
    }

    @Override protected TableMetadata provideTableMetadata(TableName tableName, ClusterName clusterName,
            Connection<HDFSClient> connection) throws ConnectorException {
        throw new UnsupportedException("Not yet supported");
    }

    @Override protected void alterCatalog(CatalogName catalogName, Map<Selector, Selector> map,
            Connection<HDFSClient> connection) throws UnsupportedException, ExecutionException {
        throw new UnsupportedException("Not yet supported");

    }

    @Override
    protected void alterTable(TableName name, com.stratio.crossdata.common.data.AlterOptions alterOptions,
            Connection<HDFSClient> connection) throws UnsupportedException, ExecutionException {
        throw new UnsupportedException("Not yet supported");

    }
}

File: connector-hdfs/src/main/scala/com.stratio.connector.hdfs/engine/HDFSMetadataEngine.scala
/*
 *
 *  Licensed to STRATIO (C) under one or more contributor license agreements.
 *  See the NOTICE file distributed with this work for additional information
 *  regarding copyright ownership. The STRATIO (C) licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License. You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied. See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package com.stratio.connector.hdfs.engine

import java.util

import com.stratio.connector.commons.{Metrics, Loggable}
import com.stratio.connector.commons.connection.{ConnectionHandler, Connection}
import com.stratio.connector.commons.engine.CommonsMetadataEngine
import com.stratio.connector.hdfs.HDFSConnector
import com.stratio.connector.hdfs.connection.{HDFSConnection, HDFSClient}
import com.stratio.crossdata.common.data.{ClusterName, CatalogName,
AlterOptions, TableName}
import com.stratio.crossdata.common.exceptions.{ExecutionException, UnsupportedException}
import com.stratio.crossdata.common.metadata.{TableMetadata, IndexMetadata,
CatalogMetadata}
import com.stratio.crossdata.common.statements.structures.Selector

/**
 * Class MetadataEngine.
 *
 * @param connectionHandler The connection handler that contains
 *                          the configuration.
 */
class HDFSMetadataEngine(connectionHandler: ConnectionHandler)
  extends CommonsMetadataEngine[HDFSClient](connectionHandler) with Loggable with Metrics{

  override def provideMetadata(
                                targetCluster: ClusterName,
                                connection: Connection[HDFSClient]): util.List[CatalogMetadata] =
    throw new UnsupportedException(s"Method provideMetadata is ${HDFSConnector.MethodNotSupported}")

  override def provideTableMetadata(
                                     tableName: TableName,
                                     targetCluster: ClusterName,
                                     connection: Connection[HDFSClient]): TableMetadata =
    throw new UnsupportedException(s"Method provideTableMetadata is ${HDFSConnector.MethodNotSupported}")

  override def provideCatalogMetadata(
                                       catalogName: CatalogName,
                                       targetCluster: ClusterName,
                                       connection: Connection[HDFSClient]): CatalogMetadata =
    throw new UnsupportedException(s"Method provideCatalogMetadata is ${HDFSConnector.MethodNotSupported}")

  /**
   * Operation that creates a directory in representation of the table.
   *
   * @param tableMetadata Metadata from the table to be created.
   * @param connection The HDFS connection.
   */
  override def createTable(
                            tableMetadata: TableMetadata,
                            connection: Connection[HDFSClient]): Unit = {
    logger.info("The Create table method has been invoked. This method doesn't do anything the table will be created when the first insert happens.")
  }


  override def createIndex(
    indexMetadata: IndexMetadata,
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method createIndex is ${HDFSConnector.MethodNotSupported}")

  override def dropCatalog(
    name: CatalogName,
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method dropCatalog is ${HDFSConnector.MethodNotSupported}")

  override def alterCatalog(
    catalogName: CatalogName,
    options: util.Map[Selector, Selector],
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method alterCatalog is ${HDFSConnector.MethodNotSupported}")

  override def alterTable(
    name: TableName,
    alterOptions: AlterOptions,
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method alterTable is ${HDFSConnector.MethodNotSupported}")

  override def dropTable(
    name: TableName,
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method dropTable is ${HDFSConnector.MethodNotSupported}")

  override def dropIndex(
    indexMetadata: IndexMetadata,
    connection: Connection[HDFSClient]): Unit =
    throw new UnsupportedException(s"Method dropIndex is ${HDFSConnector.MethodNotSupported}")

  /**
   * Operation that creates a directory in representation of the catalog.
   *
   * @param catalogMetadata Metadata from the catalog to be created.
   * @param connection The HDFS connection.
   */
  override def createCatalog(
    catalogMetadata: CatalogMetadata,
    connection: Connection[HDFSClient]): Unit = {

    logger.info("The Create catalog method has been invoked. This method doesn't do anything the table will be created when the first insert happens.")

  }
}
